File: c++/sorting/mergeSort.cpp
#include <vector>
#include <string>
using namespace std;
// Definition for a Pair
class Pair {
public:
    int key;
    string value;

    Pair(int key, string value) : key(key), value(value) {}
};

class Solution
{
public:
    vector<Pair> mergeSort(vector<Pair> &pairs)
    {
        // base case
        if (pairs.size() <= 1)
        {
            return pairs;
        }

        size_t mid = pairs.size() / 2;
        vector<Pair> arr1(pairs.begin(), pairs.begin() + mid);
        vector<Pair> arr2(pairs.begin() + mid, pairs.end());

        // recurse on both the segments
        vector<Pair> res1 = mergeSort(arr1);
        vector<Pair> res2 = mergeSort(arr2);

        // filter the two sorted arrays
        vector<Pair> res;
        res.reserve(pairs.size()); // reserve space to avoid relocation for better T
        auto it1 = res1.begin();
        auto it2 = res2.begin();

        while (it1 != res1.end() && it2 != res2.end())
        {
            if (it1->key <= it2->key)
            {
                res.push_back(*it1++);
            }
            else
            {
                res.push_back(*it2++);
            }
        }

        res.insert(res.end(), it1, res1.end()); // position, it1, it2 when inserting from something else
        res.insert(res.end(), it2, res2.end());

        return res;
    }
};

File: Python/Sorting/MergeSort.py
# Definition for a pair.
class Pair:
    def __init__(self, key: int, value: str):
        self.key = key
        self.value = value
class Solution:
    def mergeSort(self, pairs: List[Pair]) -> List[Pair]:
        if len(pairs) <= 1:
            return pairs
        
        arr1 = self.mergeSort(pairs[:(len(pairs))//2])
        arr2 = self.mergeSort(pairs[(len(pairs))//2:])

        # using two pointers
        res = []
        i, j = 0, 0
        while i < len(arr1) and j < len(arr2):
            if (arr1[i].key <= arr2[j].key):
                res.append(arr1[i])
                i += 1
                continue
            res.append(arr2[j])
            j += 1
        
        res.extend(arr1[i:])
        res.extend(arr2[j:])

        return res
    


    """
     - Time complexity: O(nlog(n)) because it is T = 2T(n/2) + n which is nlog(n)
     - Space complexity: O(n) as two arrays are made to store the data from the previous array
        The space complexity is really log(n) + n due to the size of the call stack and the arrays
        That's why it's dominated by the n and thus it is O(n).

    """



